python/ray/data/impl/arrow_block.py

```diff
@@ -265,45 +271,42 @@ def sort_and_partition(
             # so calling sort_indices() will raise an error.
             return [self._empty_table() for _ in range(len(boundaries) + 1)]
 
-        import pyarrow.compute as pac
+        context = DatasetContext.get_current()
+        col, _ = key[0]
+        if context.use_polars and pl is not None:
+            df = pl.from_arrow(self._table)
+            table = df.sort(col, reverse=descending).to_arrow()
```

Comment:
I have a general question about polars integration: IIUC, polars won't honor `num_cpus=` and will try to use as many cpus as possible. Will this be a problem?

Reply:
Yeah, it could be a problem for performance, but it shouldn't affect correctness. I expect it will still be faster than the current implementation even if there is contention.

My plan for this PR was to try benchmarking on our nightly test compute config and compare against a version where we set the number of CPUs lower to give each polars task more cores.

Reply:
I don't know at which level you set `num_cpu`, but if you can `POLARS_MAX_THREADS` before polars is imported to reduce the no. of threads.

It defaults to the number of virtual cores on the CPU.

Reply:
Ah thanks for the tip! I can try testing with that but it might be a bit tricky to ensure that the env var always gets set on all of the workers before import. Is there a way to set it after polars is initialized?

Reply:
No, once the threadpool is set, we cannot change it. :/

Reply:
Hmm yeah so unfortunately limiting the number of threads available did not change anything, but I did find that running it with 500MB partitions instead of 1GB worked in addition to limiting polars to 8 concurrent tasks at a time from the results pasted in the PR thread. I think this is more an issue with Ray's heap memory management than polars, though; we've seen the same problem in current Datasets code depending on the operation and what partition size you choose. This is an [issue](https://github.com/ray-project/ray/issues/24176) we're actively working on, so I think it's fine to merge the polars while we fix the underlying problem.